README.md
# Button-Symbols
Sketch plugin using symbol overrides to re-label, re-size, and re-position buttons. Supports nested symbols, complex button systems, icons and more!

## What can this plugin do for me?

* Use symbols for buttons.
    * Supports simple and complex button systems using various symbol replacement approaches.
	* Change the label (text override)
	* Change the horizontal and vertical padding (text overrides)
	* Change the offset position for the button (anchor to top, left, right, bottom, or center)

## What is New in 1.4.0 (replaces 1.3.0)

* Added a new command: Convert Selection to Button.
    * Covert your current selection into a valid symbol recognized by the Button Symbols plugin.
    * Shortcut: Ctrl + Cmd + Shift + B
    * Simplest usage pattern: Create a shape layer for your button background and run this command.
* Improved flexibility of symbol layer structure and naming.
    * Removed the requirement of having a layer named "Label" inside the symbol.
    * Removed the requirement of having a layer named "Background" inside the symbol.
    * Added support for a new special layer named "Label Placeholder" which was needed for the creation of some complex button systems.
* Added a Sketch file containing an advanced button system with a lot of flexibility for real world use.
    * Feel free to copy this as a starting point and customize for your own needs.
    * Learn how to <a href="https://github.com/mwhite05/Button-Symbols/blob/master/Advanced Button System.md">create your own advanced button system</a> from scratch.

## Install with Sketchpacks

<a href="https://sketchpacks.com/mwhite05/Button-Symbols/install">
  <img width="160" height="41" src="http://sketchpacks-com.s3.amazonaws.com/assets/badges/sketchpacks-badge-install.png" >
</a>

## Demo!

<p><img src="https://github.com/mwhite05/Button-Symbols/blob/master/Button Symbols 1.3.0 - Basic Demo-150.gif?raw=true" alt="Button Symbols 1.3.0 - Basic Demo"></p>

At the end of the gif I show how the position works relative to the group. The button itself is considered part of the group so if the button is outside the area you want to position inside of then you need to move the button within that area before running the plugin. Otherwise the button ends up being used to position itself which has some odd side effects.

## Usage / Instructions

The simplest way to get started with Button Symbols version 1.4.0+ is to:

* Create a rectangle for your button background.
* Create a text layer for your button label.
* Select both layers and run the "Convert Selection to Button" command (shortcut: ctrl cmd shift B)
* You now have a new symbol that is fully compatible with the Button Symbols command and an instance of that symbol is already selected.
* In the Inspector panel you will see the symbol overrides. Adjust values as desired.
* With the symbol instance layer still selected, use the `cmd + j` keyboard shortcut to resize and reposition the button. (Pro-Tip: Sketch doesn't see the updated override value until the field loses focus so be sure to tab away from the field before trying to use the shortcut!)

## Offset Container

The button position is calculated relative to its parent group. If it is not inside any group then the artboard is used. If the button is outside of all artboards then the page is used. The group|artboard|page that is used is referred to as the `Offset Container` for the remainder of this document.

## What do the Overrides Do?

* **Label (or first text layer):**
    * Change this to set a custom label for your button. This label should always be a single line of text.
    * A blank (empty) value tells the plugin to use the default label (from the master symbol).

* **Padding-H:**
    * Set this to a positive integer to use a custom amount of horizontal padding for your button.
    * The number you use will be added to the left **and** right so a value of `10` would be 10 pixels of left padding and 10 pixels of right padding.
    * Leave this value blank (empty) to have your button use the horizontal padding from the master symbol itself.
    * Use a value of `custom` to manually set the width of your button.

* **Padding-V:**
    * Set this to a positive integer to use a custom amount of vertical padding for your button.
    * The number you use will be added to the top **and** bottom so a value of `10` would be 10 pixels of top padding and 10 pixels of bottom padding.
    * Leave this value blank to use the vertical padding from the master symbol itself.
    * Use a value of `custom` to manually set the height of your button.

* **Position-X:**
    * Set this to a positive number to position the button relative to the left edge of its offset container.
    * Use a negative number to position relative to the right edge of the offset container.
    * Use a value of `center` to horizontally center the button within its offset container.
    * Leave blank to manually position the button horizontally.
    * Any label width changes made when Position-X is blank will anchor the resize from the horizontal center.
    * Use a value of `left` or `right` to position the left or right edge of the button flush with offset container's left or right edge.
    * Use a comma separated pair of integers to create a "stretchy" button. For example: `20,20` will set the button to full width with 20 pixels margin left and 20 pixels margin right. 

* **Position-Y:**
    * Set this to a positive number to position the button relative to the top edge of its offset container.
    * Use a negative number to position relative to the bottom edge of the offset container.
    * Use a value of `center` to vertically center the button within its offset container.
    * Leave blank to manually position the button vertically.
    * Any label width changes made when Position-Y is blank will anchor the resize from the vertical center.
    * Use a value of `top` or `bottom` to position the top or bottom edge of the button flush with offset container's top or bottom edge.
    * Use a comma separated pair of integers to create a "stretchy" button. For example: `20,20` will set the button to full height with 20 pixels margin top and 20 pixels margin bottom.

## Future Plans

* Handling of events/Actions: The TextChanged event is only run when the text layer value is changed directly and does not run when a symbol override changes. I'm currently looking into alternative solutions that will work via the Cocoascript API. At this time I have no plans to rewrite this plugin in native code.
* [Use the issue reports page to request features](https://github.com/mwhite05/Button-Symbols/issues). No promises that I will add them, but feel free to ask.



Changelog.md
# Changelog

## v1.4.0
**Date:** 2018-03-16

1. THIS IS BIG! Added an Sketch file and tutorial for an advanced button system with a lot of flexibility for real world use!
1. Added support for more flexible and advanced layer structures inside of a button symbol.
    a. Removed the requirement of having a layer named Background.
    b. Removed the requirement of having a text layer named Label (as long as you provide a Label Placeholder shape layer instead).
    c. Added support for a new Label Placeholder layer for providing the plugin with Label size information when the Label layer itself is not present or is named something other than Label.
    d. Removed the requirement of
3. Updated the utility class console logging to send output to the log() function if available (instead of console.log)

## v1.3.0
**Date:** 2017-09-01

1. Added support for `custom` as a special padding value. This allows you to take advantage of the position and label override features while controlling the button's width and/or height manually.
2. Added support for "stretchy" buttons by allowing a comma separated list of offset values (e.g. Position-X: &lt;left&gt;,&lt;right&gt; or Position-Y: &lt;top&gt;,&lt;bottom&gt;) If
3. Code cleanup to divide it into multiple files for better organization and easier maintenance of this plugin.
4. Improved the bundled symbol in regards to distribution to the world by changing the font in the default button to use Verdana instead of Open Sans and adjusted the name of the bundled symbol to "ButtonSymbols/Primary" (instead of just "Button").
5. Updated the demo on the readme page.
6. Updated the documentation of the available features.

## v1.2.4-alpha.1
**Date:** 2017-08-30

1. Added a new Symbol.sketch file as a resource that the plugin can use to insert a properly configured button master symbol.
2. Added a demo gif to the readme page.
3. Improved readme by specifying what type of layer each special layer in the button symbol must be.
4. Syntax correction where variables were implicitly declared.
5. Added a new idea to the future plans section of the readme.

## v1.2.3
**Date:** 2017-08-29

1. Much headdeskery. Patched an issue where the install link did not actually point to this plugin.

## v1.2.2
**Date:** 2017-08-29

1. Updated the readme to better outline this plugin when viewed in the Sketchpacks installer (only the first sentence of the readme file shows up there).
2. Added the Install with Sketchpacks badge to the readme file. 

## v1.2.1
**Date:** 2017-08-29

1. Added support for Sketchpacks in the manifest.json configuration file.

## v1.2.0
**Date:** 2017-08-28

1. Updated the plugin for Sketch 44.0+ by patching an issue where the layer overrides are no longer returned as an object wrapped by an array. They are now returned as just the object itself.

## v1.1.1
**Date:** 2017-04-22

Patch release.
 
1. Fixed an issue where the Position-X value when set to 'center' would cause it to behave as if the Position-Y value were also set to 'center'.

## v1.1.0
**Date:** 2017-04-21

Updated release to support some additional word based position options. Until this release it was not possible to have a button automatically position flush with the edges of the parent container.

1. Use `left` or `right` for `Position-X` to make button flush to left or right edge.
2. Use `top` or `bottom` for `Position-Y` to make button flush to top or bottom edge.

## v1.0.0
**Date:** 2017-04-18

Initial release of functional plugin capable of sizing and positioning a button using a symbol instance.

1. Use a symbol as a button.
2. Use special layer names in the button symbol to allow setting overrides for:
    * Label
    * Padding-H
    * Padding-V
    * Position-X
    * Position-Y
3. Keyboard shortcut `cmd + j` to run resize/reposition command on selected layers.
4. Use `center` for position values to center the button.
5. Positioning works to parent container (i.e. group, artboard).



TestScript.js
/*

    Debugging plugins via the Safari console is a pain because each run opens a new console window.

    Open the Plugins > Run Script menu in Sketch and paste in the script below. By using this you can call any ButtonSymbols method directly to test it.

    Just use clog() anywhere in the plugin to log to the Sketch plugin console if available; otherwise it will log to the Safari console if available.


    - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - 

    Useful tip: Reloading sketch plugins every time you change something is painful.

    Navigate to: /Users/mwhite/Library/Application Support/com.bohemiancoding.sketch3/Plugins/Button Symbols.sketchplugin/Contents/Sketch

    Remove the button-symbols.js file and create a symlink named button-symbols.js that points to the copy of that file inside the repository.
    Also do this for any other files being worked on.

*/

const sketch = require('sketch');

@import 'Button Symbols.sketchplugin/Contents/Sketch/mwhite.base-class.js'
@import 'Button Symbols.sketchplugin/Contents/Sketch/mwhite.util.js'
@import 'Button Symbols.sketchplugin/Contents/Sketch/button-symbols.js'

var bs = new ButtonSymbols(context);
bs.updateSelectedButtons();


Button Symbols.sketchplugin/Contents/Sketch/mwhite.util.js
/**
 * @author Michael White
 * @copyright 2017 Michael White
 */

if(typeof(Mwhite) == 'undefined' || !(Mwhite instanceof Object)) {
    Mwhite = {};
}

Mwhite.Util = {
    _getClogs: function(args) {
        var clogs = [];
        var argsLength = args.length;
        if(argsLength && ['function', 'object'].indexOf(typeof(console.log)) >= 0) {
            var scalarTypes = ['string', 'boolean', 'number', 'undefined'];
            var previousClogItemType = '';
            for(var i = argsLength; i >= 1; i--) {
                var clogItem = args[argsLength - i];
                var clogItemType = typeof(clogItem);
                // Is scalar?
                if(scalarTypes.indexOf(clogItemType) >= 0 && scalarTypes.indexOf(previousClogItemType) >= 0) {
                    // Concat with previous clog item
                    if(clogItemType == 'boolean') {
                        clogItem = '(' + (clogItem ? 'true' : 'false') + ')';
                    } else if (clogItemType == 'undefined') {
                        clogItem = '(undefined)';
                    }
                    clogs[clogs.length - 1] += ': ' + clogItem;
                } else {
                    clogs.push(clogItem);
                }
                previousClogItemType = clogItemType;
            }
        }
        return clogs;
    },
    clog: function() {
        var args = Array.prototype.slice.call(arguments);
        var clogs = Mwhite.Util._getClogs(args);
        for(var i in clogs) {
            if(clogs.hasOwnProperty(i)) {
                var clogItem = clogs[i];
                if(typeof log === 'function') {
                    log(clogItem);
                } else {
                    console.log(clogItem);
                }
            }
        }
    }
};

if(typeof(clog) != 'function') {
    clog = Mwhite.Util.clog;
}


Button Symbols.sketchplugin/Contents/Sketch/action-handlers.js
/**
 * @author Michael White
 * @copyright 2017 Michael White
 */

@import "button-symbols.js";

/**
 * Not really used for now. I hope to use this in the future.
 *
 * @param context
 */
function onSelectionChanged(context) {
    clog('running onSelectionChanged()');
    // Do nothing for now - maybe we automate this later
    return;
    /*var action = context.actionContext;
     var doc = action.document;
     var layers = action.newSelection;
     _updateButtonSymbols(layers);*/
}

/**
 * Not really used for now. I hope to use this in the future.
 *
 * @param context
 */
function onTextChanged(context) {
    clog('running onTextChanged()');
    // todo - once this method will actually run we can automate it to update the button symbol the text is part of.
    return;
    /*var action = context.actionContext;
     var doc = action.document;
     doc.showMessage('Text changed in document.');*/
}


Button Symbols.sketchplugin/Contents/Sketch/plugin.js
/**
 * @author Michael White
 * @copyright 2017 Freckle, LLC
 */

function clog() {
    var args = Array.prototype.slice.call(arguments);
    var clogs = _getClogs(args);
    for(var i in clogs) {
        if(clogs.hasOwnProperty(i)) {
            var clogItem = clogs[i];
            console.log(clogItem);
        }
    }
}

function _getClogs(args) {
    var clogs = [];
    var argsLength = args.length;
    if(argsLength && ['function', 'object'].indexOf(typeof(console.log)) >= 0) {
        var scalarTypes = ['string', 'boolean', 'number', 'undefined'];
        var previousClogItemType = '';
        for(var i = argsLength; i >= 1; i--) {
            var clogItem = args[argsLength - i];
            var clogItemType = typeof(clogItem);
            // Is scalar?
            if(scalarTypes.indexOf(clogItemType) >= 0 && scalarTypes.indexOf(previousClogItemType) >= 0) {
                // Concat with previous clog item
                if(clogItemType == 'boolean') {
                    clogItem = '(' + (clogItem ? 'true' : 'false') + ')';
                } else if (clogItemType == 'undefined') {
                    clogItem = '(undefined)';
                }
                clogs[clogs.length - 1] += ': ' + clogItem;
            } else {
                clogs.push(clogItem);
            }
            previousClogItemType = clogItemType;
        }
    }
    return clogs;
}

function _insertButtonSymbol(context) {
    var layers = context.selection;
    clog('_insertButtonSymbol()');
    clog(layers.count());
    for(var k = 0; k < layers.count(); k++) {
        //reference each layer of each artboard
        var layer = layers[k];
        clog('Layer Number', k+1);
        clog(layer.class() + '');
        clog(layer.name() + '');

        clog('is artboard', isArtboard(layer));
        clog('is group', isGroup(layer));
        clog('is text', isText(layer));

        var group = null;
        if(isText(layer)) {
            // The idea here is to get the text layer's value and then replace the text layer with a button instance that contains the text layer's value as its label override.
                /*
                 *
                 * var obj = {};
    obj[<overrideLayerID>] = "overrideText";
    symbolInstance.addOverrides_forCellAtIndex_ancestorIDs_(obj, 0, nil);
                 */
        } else {
            if(isGroup(layer)) {
                clog('GROUP FOUND');
                group = layer;
            } else {
                clog('FINDING PARENT GROUP');
                group = layer.parentGroup();
            }

            // Insert the symbol into the target group.
            var sketch = context.api();
            var symbolUri = sketch.resourceNamed('Symbol.sketch');


	        var sketchFile = MSDocument.new();
	        sketchFile.readFromURL_ofType_error(symbolUri, 'com.bohemiancoding.sketch.drawing', null);

            if(sketchFile) {
                _addSymbolByName(context, sketchFile, group, 'Button');
                sketchFile.close();
            } else {
                context.document.showMessage('Unable to open the Symbol.sketch file in the plugin Resources folder. Try re-installing Button Symbols to repair the plugin.');
            }
        }
    }
}

function _addSymbolByName(context, sourceDocument, targetGroup, symbolName) {
    var sourceSymbols = sourceDocument.documentData().allSymbols();
    var sourceSymbol = _getSymbolByName('Button', sourceSymbols);

    if(sourceSymbol) {
        var previousPage = _goToSymbolsPage(context);
        var sourceSymbolName = sourceSymbol.name() + '';
        // Check that the symbol by the same name doesn't already exist
        var allSymbols = context.document.documentData().allSymbols();
        var tmpSourceSymbol = _getSymbolByName(sourceSymbolName, allSymbols);
        if(!tmpSourceSymbol) {
            context.document.currentPage().addLayers([sourceSymbol]);
            allSymbols = context.document.documentData().allSymbols();
            sourceSymbol = _getSymbolByName(sourceSymbolName, allSymbols);
        } else {
            sourceSymbol = tmpSourceSymbol;
        }
        context.document.setCurrentPage(previousPage);
        // Add symbol instance to the target group
        // todo change how X,Y coordinates are determined.
        //var rect = new Rectangle(0, 0, sourceSymbol.frame().width(), sourceSymbol.frame().height());
        /*var rect = CGRectMake(0, 0, sourceSymbol.frame().width(), sourceSymbol.frame().height());
        var symbolInstance = MSSymbolCreator.createSymbolInstanceWithName_rect_symbolID(sourceSymbolName, rect, sourceSymbol.objectID());
        sourceSymbol.registerInstance(symbolInstance);*/
        // wow... it took WAY TOO LONG to find this: MSSymbolMaster.newSymbolInstance()
        var symbolInstance = sourceSymbol.newSymbolInstance();
        targetGroup.addLayers([symbolInstance]);
    }
}

/**
 * Get a symbol based on its name from a list of provided symbol layers.
 *
 * @param {String} targetSymbolName
 * @param {NSArray} symbols
 * @return {MSSymbolMaster|null}
 */
function _getSymbolByName(targetSymbolName, symbols) {
    for(var i = 0; i < symbols.count(); i++) {
        var tempSymbolName = symbols.objectAtIndex(i).name();
        if(tempSymbolName && tempSymbolName.isEqualToString(targetSymbolName)) {
            return symbols.objectAtIndex(i);
        }
    }

    return null;
}


/**
 * Update the size and position of any selected button symbols.
 *
 * @param layers A Sketch Layers object. (NOT AN ARRAY)
 * @returns {boolean}
 * @private
 */
function _updateButtonSymbols(layers) {
    for(var k = 0; k < layers.count(); k++) {
        //reference each layer of each artboard
        var layer = layers[k];

        // Check to see if the layer is a Symbol (this operation only works for symbols)
        var symbol = getSymbolForLayer(layer);
        if(symbol) {
            var symbolMaster = symbol.symbolMaster();
            var masterChildren = symbolMaster.children();
            var masterChildrenCount = masterChildren.count();

            // 2017-28-08 :: After waiting a while to fix this, turns out that layer.overrides() now returns just the overrides for this layer vs how it returned them in Sketch 43.1
            var overrides = layer.overrides();

            if(!overrides) {
                continue; // No point continuing - no overrides set
            }

            // NOTE: If an offset is positive the offset anchor is either top or left. If an offset is negative the offset anchor is bottom or right.
            var positionX = 0;
            var positionY = 0;
            var paddingH = 0;
            var paddingV = 0;
            var labelLayer = null;
            var labelTextOverride = '';
            var templateLabelWidth = 0;
            var templateLabelHeight = 0;
            var templateBackgroundWidth = 0;
            var templateBackgroundHeight = 0;

            // Loop the children of the symbol to find the layers that provide the override values we want.
            for(var i = 0; i < masterChildrenCount; i++) {
                var childLayer = masterChildren[i];
                var childName = childLayer.name() + '';
                var childObjectId = childLayer.objectID() + '';
                switch(childName) {
                    case 'Label':
                        templateLabelWidth = childLayer.frame().width();
                        templateLabelHeight = childLayer.frame().height();
                        labelLayer = childLayer;
                        labelTextOverride = overrides[childObjectId];
                        break;
                    case 'Position-X':
                        if(overrides[childObjectId]) {
                            positionX = overrides[childObjectId];
                        }
                        if(!positionX) {
                            if(childLayer.class() == 'MSTextLayer') {
                                positionX = childLayer.stringValue();
                            }
                        }
                        break;
                    case 'Position-Y':
                        if(overrides[childObjectId]) {
                            positionY = overrides[childObjectId];
                        }
                        if(!positionY) {
                            if(childLayer.class() == 'MSTextLayer') {
                                positionY = childLayer.stringValue();
                            }
                        }
                        break;
                    case 'Padding-H':
                        if(overrides[childObjectId]) {
                            paddingH = parseInt(overrides[childObjectId], 10);
                        }
                        if(!paddingH) {
                            if(childLayer.class() == 'MSTextLayer') {
                                paddingH = parseInt(childLayer.stringValue().trim(), 10);
                            }
                        }
                        break;
                    case 'Padding-V':
                        if(overrides[childObjectId]) {
                            paddingV = parseInt(overrides[childObjectId], 10);
                        }
                        if(!paddingV) {
                            if(childLayer.class() == 'MSTextLayer') {
                                paddingV = parseInt(childLayer.stringValue().trim(), 10);
                            }
                        }
                        break;
                    case 'Background':
                        templateBackgroundWidth = childLayer.frame().width();
                        templateBackgroundHeight = childLayer.frame().height();
                        break;
                }
            }

            // Store the master symbol's label text
            var defaultSymbolTextLayerValue = labelLayer.stringValue();
            if(labelTextOverride == '' || labelTextOverride == null) {
                labelTextOverride = defaultSymbolTextLayerValue;
            }

            if(!paddingH) {
                // No default or override horizontal padding. Calculate it from existing dimensions in the master symbol.
                paddingH = (templateBackgroundWidth - templateLabelWidth) / 2;
            }

            if(!paddingV) {
                // No default or override vertical padding. Calculate it from existing dimensions in the master symbol.
                paddingV = (templateBackgroundHeight - templateLabelHeight) / 2;
            }

            // If no existing x or y position is set, treat the center position of this symbol as the anchor point for the resize
            var maintainCenterH = false;
            var maintainCenterV = false;
            var centerToParentH = false;
            var centerToParentV = false;
            positionX = positionX.toString().trim();
            positionY = positionY.toString().trim();
            switch(positionX) {
                case 'center':
                    centerToParentH = true;
                    break;
                case 'left':
                case 'right':
                    break;
                default:
                    positionX = parseInt(positionX, 10);
                    if(!positionX) {
                        maintainCenterH = true;
                    }
                    break;
            }
            switch(positionY) {
                case 'center':
                    centerToParentV = true;
                    break;
                case 'top':
                case 'bottom':
                    break;
                default:
                    positionY = parseInt(positionY, 10);
                    if(!positionY) {
                        maintainCenterV = true;
                    }
                    break;
            }

            var currentSymbolWidth = symbol.frame().width();
            var currentSymbolHeight = symbol.frame().height();

            var currentSymbolX = symbol.frame().x();
            var currentSymbolY = symbol.frame().y();

            // Set the master symbol's label text to the value of the override text
            labelLayer.setStringValue(labelTextOverride);

            // Get the updated width and height of the label text
            var labelWidth = labelLayer.frame().width();
            var labelHeight = labelLayer.frame().height();

            // Restore the master symbol's label text to its initial value
            labelLayer.setStringValue(defaultSymbolTextLayerValue);

            var newSymbolWidth = labelWidth + (paddingH * 2);
            var newSymbolHeight = labelHeight + (paddingV * 2);

            symbol.frame().setWidth(newSymbolWidth);
            symbol.frame().setHeight(newSymbolHeight);

            // Get the container for this symbol
            var group = symbol.parentGroup();

            // Reposition the button within its container as required
            if(!group) {
                return false;
            }

            if(maintainCenterH) {
                // Figure out the left offset as being changed by 1/2 of the change in button width
                positionX = currentSymbolX - ((newSymbolWidth - currentSymbolWidth) / 2);
            } else if(centerToParentH) {
                // Find half the width of the container and subtract half the width of the button
                positionX = (group.frame().width() / 2) - (newSymbolWidth / 2);
            } else {
                switch(positionX) {
                    case 'left':
                        // Position flush with left edge of parent container
                        positionX = 0;
                        break;
                    case 'right':
                        // Use (width of container - button width) to position button flush to right edge of parent container
                        positionX = group.frame().width() - newSymbolWidth;
                        break;
                    default:
                        if(positionX > 0) {
                            // Offset is from left - do nothing special because this is the standard anchor point Sketch uses on resize
                        } else if(positionX < 0) {
                            // Offset is from right; adjust based on current coordinate space
                            positionX = (group.frame().width() - newSymbolWidth) + positionX;
                        }
                        break;
                }
            }

            if(maintainCenterV) {
                // Figure out the left offset as being changed by 1/2 of the change in button width
                positionY = currentSymbolY - ((newSymbolHeight - currentSymbolHeight) / 2);
            } else if(centerToParentV) {
                // Find half the height of the container and subtract half the height of the button
                positionY = (group.frame().height() / 2) - (newSymbolHeight / 2);
            } else {
                switch(positionY) {
                    case 'top':
                        // Position flush with top of parent container
                        positionY = 0;
                        break;
                    case 'bottom':
                        // Use (height of container - button height) to position button flush to bottom of parent container
                        positionY = group.frame().height() - newSymbolHeight;
                        break;
                    default:
                        if(positionY > 0) {
                            // Offset is from top - do nothing special because this is the standard anchor point Sketch uses on resize
                        } else if(positionY < 0) {
                            // Offset is from bottom ; adjust based on current coordinate space
                            positionY = (group.frame().height() - newSymbolHeight) + positionY;
                        }
                        break;
                }
            }

            symbol.frame().setX(positionX);
            symbol.frame().setY(positionY);
        }
    }
}

/**
 * This is the method called when the user runs the plugin from the menu or keyboard shortcut.
 *
 * @param context
 */
function command_updateButtonSymbols(context) {
    var layers = context.selection;
    _updateButtonSymbols(layers);
}

function command_insertButtonSymbol(context) {
    _insertButtonSymbol(context);
}

/**
 * Not really used for now. I hope to use this in the future.
 *
 * @param context
 */
function onSelectionChanged(context) {
    //clog('running onSelectionChanged()');
    // Do nothing for now - maybe we automate this later
    return;
    /*var action = context.actionContext;
     var doc = action.document;
     var layers = action.newSelection;
     _updateButtonSymbols(layers);*/
}

/**
 * Not really used for now. I hope to use this in the future.
 *
 * @param context
 */
function onTextChanged(context) {
    //clog('running onTextChanged()');
    // todo - once this method will actually run we can automate it to update the button symbol the text is part of.
    return;
    /*var action = context.actionContext;
     var doc = action.document;
     doc.showMessage('Text changed in document.');*/
}

function _goToSymbolsPage(context) {
    var previousPage = context.document.currentPage();
    context.document.setCurrentPage(context.document.documentData().symbolsPageOrCreateIfNecessary());
    return previousPage;
}

/**
 * Utility to determine whether the specified layer is a symbol of any kind (master or instance)
 *
 * @param {MSLayer} layer
 * @returns {boolean}
 */
function isSymbol(layer) {
    var className = layer.class() + '';
    return className == 'MSSymbolMaster' || className == 'MSSymbolInstance';
}

/**
 * Utility to determine whether the specified layer is a symbol instance.
 *
 * @param {MSLayer} layer
 * @returns {boolean}
 */
function isSymbolInstance(layer) {
    var className = layer.class() + '';
    return className == 'MSSymbolInstance';
}

/**
 * Utility to determine whether the specified layer is a master symbol.
 *
 * @param {MSLayer} layer
 * @returns {boolean}
 */
function isMasterSymbol(layer) {
    var className = layer.class() + '';
    return className == 'MSSymbolMaster';
}

/**
 * Utility to determine whether the specified layer is an artboard.
 *
 * @param {MSLayer} layer
 * @returns {boolean}
 */
function isArtboard(layer) {
    var className = layer.class() + '';
    return className == 'MSArtboardGroup';
}

/**
 * Utility to determine whether the specified layer is a group.
 *
 * @param {MSLayer} layer
 * @returns {boolean}
 */
function isGroup(layer) {
    var className = layer.class() + '';
    return className == 'MSGroupLayer' || className == 'MSArtboardGroup';
}

/**
 * Utility to determine whether the specified layer is a text layer.
 *
 * @param {MSLayer} layer
 * @returns {boolean}
 */
function isText(layer) {
    var className = layer.class() + '';
    return className == 'MSTextLayer';
}

/**
 * Utility to determine whether the specified layer is a group layer.
 *
 * @param {MSLayer} layer
 * @returns {boolean}
 */
function isImage(layer) {
    var className = layer.class() + '';
    return className == 'MSBitmapLayer';
}

/**
 * Utility to determine whether the specified layer is a shape layer.
 *
 * @param {MSLayer} layer
 * @returns {boolean}
 */
function isShape(layer) {
    var className = layer.class() + '';
    return className == 'MSShapeGroup';
}

/**
 * Utility to get the symbol that the specified layer is part of. If the layer is not in a symbol null is returned.
 *
 * @param layer
 * @returns {*}
 */
function getSymbolForLayer(layer) {
    if(layer) {
        if(isSymbol(layer)) {
            return layer;
        }

        return getSymbolForLayer(layer.parentGroup());
    }

    return null;
}

// Code for debugging as Sketch script (vs installed plugin)
//var selectedLayers = context.selection;
//_updateButtonSymbols(selectedLayers);
